### PR TITLE
Full type hinting

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
A follow up PR to https://github.com/openai/gym/pull/2939 that was inspired by https://github.com/openai/gym/issues/2918 to fully type check the whole gym project (along with tests).
This is removes all files and folders from the exlude folder however more work is required to add the core API to the strict parameter. 
However, this should catch any weird or incorrect type hints but at this time not force PRs to include type hints

In the process, I discovered a couple of strange tests for async vector envs which needs fixing however it was outside the scope of this PR

### CODE REVIEW

File: gym/utils/play.py [hunk 1]
```diff
@@ -120,11 +129,11 @@ def display_arr(
 
 def play(
     env: Env,
+    keys_to_action: Dict[Union[Tuple[Union[str, int]], str], ActType],
```
Comment (RedTachyon) [inline gym/utils/play.py:132]: This seems like an unnecessary breaking change? (changing the order of arguments)

File: gym/vector/utils/__init__.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
     read_from_shared_memory,
     write_to_shared_memory,
 )
-from gym.vector.utils.spaces import BaseGymSpaces, _BaseGymSpaces, batch_space, iterate
```
Comment (RedTachyon) [inline gym/vector/utils/__init__.py:9]: I kept this export deliberately. It was a bad idea and shouldn't be made private initially, but it's not actively harmful. One less breaking change for people to complain about.

File: gym/vector/vector_env.py [hunk 3]
```diff
@@ -69,7 +69,7 @@ def reset_wait(
         seed: Optional[Union[int, List[int]]] = None,
         return_info: bool = False,
         options: Optional[dict] = None,
-    ):
+    ) -> np.ndarray:
```
Comment (RedTachyon) [inline gym/vector/vector_env.py:72]: Don't vector envs have the same (obs, info) semantic as general envs?

File: gym/wrappers/gray_scale_observation.py [hunk 4]
```diff
@@ -32,11 +32,12 @@ def __init__(self, env: gym.Env, keep_dim: bool = False):
...
 
         assert (
-            len(env.observation_space.shape) == 3
+            isinstance(env.observation_space, Box)
+            and len(env.observation_space.shape) == 3
             and env.observation_space.shape[-1] == 3
         )
 
-        obs_shape = self.observation_space.shape[:2]
+        obs_shape = env.observation_space.shape[:2]
```
Comment (RedTachyon) [inline gym/wrappers/gray_scale_observation.py:40]: This looks like it could change the behavior of the code, why is this necessary?

File: gym/vector/vector_env.py (hunk 3 above)
Comment (pseudo-rnd-thoughts) [reply gym/vector/vector_env.py:72]: I thought it was the cause of some issues but it doesn't raise an error anymore so will revert

File: gym/vector/utils/__init__.py (hunk 2 above)
Comment (pseudo-rnd-thoughts) [reply gym/vector/utils/__init__.py:9]: This is in `vector/utils/__init__.py` so if someone is getting this from here then there is an issue

File: gym/wrappers/normalize.py [hunk 5]
```diff
@@ -81,19 +81,20 @@ def step(self, action):
 
     def reset(self, **kwargs):
         """Resets the environment and normalizes the observation."""
-        return_info = kwargs.get("return_info", False)
-        if return_info:
+        if kwargs.get("return_info", False):
             obs, info = self.env.reset(**kwargs)
+
+            if self.is_vector_env:
```
Comment (RedTachyon) [inline gym/wrappers/normalize.py:87]: This also looks like a change of behavior, is this intended? Can you explain it?

File: gym/wrappers/resize_observation.py [hunk 6]
```diff
@@ -39,7 +39,16 @@ def __init__(self, env: gym.Env, shape: Union[tuple, int]):
 
         self.shape = tuple(shape)
 
-        obs_shape = self.shape + self.observation_space.shape[2:]
+        assert isinstance(
+            env.observation_space, Box
+        ), f"Expected the observation space to be Box, actual type: {type(env.observation_space)}"
+        assert len(env.observation_space.shape) == 3 and env.observation_space.shape[
```
Comment (RedTachyon) [inline gym/wrappers/resize_observation.py:45]: This seems irrelevant to typing, right? Is this actually necessary for this wrapper to work, or is there any conceivable situation where this could break otherwise functional code?

File: gym/wrappers/gray_scale_observation.py (hunk 4 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/gray_scale_observation.py:40]: Good point, changed back. The assert statements fixed the problem

File: gym/utils/play.py (hunk 1 above)
Comment (pseudo-rnd-thoughts) [reply gym/utils/play.py:132]: I spoke to @Markus28 about it. This was a bug in the previous implementation as `keys_to_action` is a required parameter if you look at the for loop after the docstring.

File: tests/vector/test_async_vector_env.py [hunk 7]
```diff
@@ -28,11 +28,11 @@ def test_create_async_vector_env(shared_memory):
 @pytest.mark.parametrize("shared_memory", [True, False])
 def test_reset_async_vector_env(shared_memory):
     env_fns = [make_env("CartPole-v1", i) for i in range(8)]
-    try:
-        env = AsyncVectorEnv(env_fns, shared_memory=shared_memory)
-        observations = env.reset()
-    finally:
-        env.close()
+
+    env = AsyncVectorEnv(env_fns, shared_memory=shared_memory)
```
Comment (RedTachyon) [inline tests/vector/test_async_vector_env.py:32]: This is one of the situations where it might be better to do a pyright ignore, or maybe something more clever.

Basically what might (maybe?) happen here is that something crashes inside the environment, the test gets aborted, and then the environment is never properly closed, which can lead to a memory leak, because there are some hanging processes. The `finally` clause prevents that.

It might be possible by using a feature that's existed in gym very quietly, which is using an env as a context manager:
```python
with AsyncVectorEnv(env_fns, shared_memory=shared_memory) as env:
    ... # do whatever
```
I'm not 100% sure, but I think this will make sure to close the environment even if it crashes

Comment (RedTachyon) [reply tests/vector/test_async_vector_env.py:32]: (on second thought, I'm not sure if this is implemented for VectorEnv, but it definitely is for normal envs)

File: gym/wrappers/normalize.py (hunk 5 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/normalize.py:87]: I believe that it is not a change in behaviour. The problem previous was that `info` could have been unbounded. To avoid the issue, I have all of the return info code in one part of the if statement and without info is the second part. This has the cost of copying the if vector code twice but it shouldn't be a issue

File: tests/vector/test_spaces.py [hunk 8]
```diff
@@ -116,9 +116,11 @@ def test_batch_space_custom_space(space, expected_batch_space_4):
 def test_iterate(space, batch_space):
     items = batch_space.sample()
     iterator = iterate(batch_space, items)
+    final_i = 0
```
Comment (RedTachyon) [inline tests/vector/test_spaces.py:119]: Just do `i=0` before the loop and get rid of `final_i`

File: gym/wrappers/resize_observation.py (hunk 6 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/resize_observation.py:45]: You right I will remove, the assert is not necessary for the PR but I think true however could cause more issues

File: gym/utils/play.py (hunk 1 above)
Comment (Markus28) [reply gym/utils/play.py:132]: I think I would actually prefer to leave `key_to_action` optional and try to fetch it from the environment, then throw an exception if that's not possible. That's how it used to work (before I kinda unknowingly broke it with the for loop).

Comment (Markus28) [reply gym/utils/play.py:132]: Also, I need to update the tests for utils.play, in particular they currently don't seem to cover the case where `env` provides keys_to_action.

File: tests/vector/test_spaces.py (hunk 8 above)
Comment (pseudo-rnd-thoughts) [reply tests/vector/test_spaces.py:119]: Done

File: gym/utils/play.py (hunk 1 above)
Comment (pseudo-rnd-thoughts) [reply gym/utils/play.py:132]: Im going to remove `utils.play` from this PR and let @Markus28 fix it in another PR

File: tests/envs/utils.py [hunk 9]
```diff
@@ -18,22 +18,27 @@ def try_make_env(env_spec: EnvSpec) -> Optional[gym.Env]:
...
     for env_spec in all_testing_env_specs
     if "gym.envs.mujoco" in env_spec.entry_point
 ]
-gym_testing_env_specs = [
+gym_testing_env_specs: List[EnvSpec] = [
     env_spec
     for env_spec in all_testing_env_specs
     if any(
-        f"gym.{ep}" in env_spec.entry_point
+        f"gym.envs.{ep}" in env_spec.entry_point
```
Comment (RedTachyon) [inline tests/envs/utils.py:41]: Why was this changed? Is this a bugfix unrelated to the typing?

Comment (pseudo-rnd-thoughts) [reply tests/envs/utils.py:41]: It is unrelated to type hinting but it fixes a previous bug
